### PR TITLE
Clean up user selectors

### PR DESCRIPTION
**Pull request by kulyk**
Started with an indent to simplify null checks in `metabase/selectors/user`; ended up adding missing types and converting the whole thing to TypeScript. Had to patch `metabase/plugins` to use the `User` type coming from `metabase-types/api` vs. deprecated `metabase-types/types`; also cleaned up the imports there as well. No changes in behavior are expected.

### CODE REVIEW

**Review by ranquild — APPROVED** [review]
@kulyk WDYT about changing `currentUser: User;` to `currentUser: User | null;` in `State`? This matches https://github.com/metabase/metabase/blob/2d1b9801c0b4e6ba46baa073efb87ce57aeee21d/frontend/src/metabase/redux/user.js#L52 where the default reducer value is `null`, and it will be `null` until the user is loaded. Currently it's unclear why we have `?` checks in selectors just by looking at the type.

File: frontend/src/metabase/selectors/user.ts [hunk 1]
```diff
@@ -21,10 +23,10 @@ export const canManageSubscriptions = createSelector(
...
 export const getUserAttributes = createSelector(
   [getUser],
-  user => (user && user.login_attributes) || [],
+  user => user?.login_attributes || [],
 );
 
 export const getUserPersonalCollectionId = createSelector(
   [getUser],
-  user => (user && user.personal_collection_id) || null,
+  user => user?.personal_collection_id || null,
```
Comment (ranquild) [inline frontend/src/metabase/selectors/user.ts:31]: With correct types I believe that `|| null` is not needed. `state.currentUser` can be `User | null`, so `state.currentUser?.personal_collection_id` will return either `null` or `personal_collection_id` which is `number`.

Comment (ranquild) [reply frontend/src/metabase/selectors/user.ts:31]: But I understand that it might be safer to keep this check

**Comment by ranquild** [comment]
@kulyk `checkNotNull` is merged and can be used here

File: frontend/src/metabase/plugins/index.ts [hunk 2]
```diff
@@ -74,7 +78,7 @@ export const PLUGIN_AUTH_PROVIDERS = [] as any;
 
 // Only show the password tab in account settings if these functions all return true.
 // Otherwise, the user is logged in via SSO and should hide first name, last name, and email field in profile settings metabase#23298.
-export const PLUGIN_IS_PASSWORD_USER: ((user: User) => boolean)[] = [];
+export const PLUGIN_IS_PASSWORD_USER: ((user: User | null) => boolean)[] = [];
```
Comment (WiNloSt) [inline frontend/src/metabase/plugins/index.ts:81]: Can we still keep the signature as is? If you look at some of the usage
https://github.com/metabase/metabase/blob/c622038e34947dedbb6978b87b6ebab346454825/frontend/src/metabase/plugins/builtin/auth/jwt.js#L3

It's clear that that function doesn't expect a `user` to be nullable. If this causes error somewhere else you could cast the argument to be `User` in the same spirit as other files in this PR.

https://github.com/metabase/metabase/pull/26477/files#diff-4aeb510eee744ac8cce724e9c015559b0fade8862f9812ef084ef7bc0688d61aR10

**Review by WiNloSt — APPROVED** [review]
I've commented on `PLUGIN_IS_PASSWORD_USER`'s type. But other than that, all looks good to me.

**Comment by kulyk** [comment]
@ranquild @WiNloSt, thanks for your reviews 🙏 I just pushed [`ced01a3`](https://github.com/metabase/metabase/pull/26477/commits/ced01a3ba467969037007eb0feea8aab1f43c440) that makes it use `checkNotNull` helper for `user` objects here and there. Could you please give it one more look? One thing I'd especially appreciate your take on is if we should handle a possible `TypeError` that `checkNotNull` can throw?

**Comment by WiNloSt** [comment]
As for
> One thing I'd especially appreciate your take on is if we should handle a possible TypeError that checkNotNull can throw?

I don't think we need to handle it. Assuming `checkNotNull` is meant to be for TypeScript and all argument pass to it should never be null.

Another angle to look at this is the previous code you change doesn't handle `null` so the new code shouldn't as `null` isn't expected. If something is `null` that means there's a bug in our application.

File: frontend/src/metabase/account/password/containers/UserPasswordApp/UserPasswordApp.tsx [hunk 3]
```diff
@@ -1,11 +1,17 @@
...
+
+import type { User } from "metabase-types/api";
+import type { State } from "metabase-types/store";
+
 import { updatePassword, validatePassword } from "../../actions";
 import UserPasswordForm from "../../components/UserPasswordForm";
 
 const mapStateToProps = (state: State) => ({
-  user: getUser(state),
+  user: checkNotNull<User>(getUser(state)),
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/account/password/containers/UserPasswordApp/UserPasswordApp.tsx:14]: @kulyk I've just tried and it works without specifying the type `<User>`. Could you remove it?

File: frontend/src/metabase/account/profile/containers/UserProfileApp/UserProfileApp.tsx [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
+import type { User } from "metabase-types/api";
+import type { State } from "metabase-types/store";
+
 import UserProfileForm from "../../components/UserProfileForm";
 import { updateUser } from "../../actions";
 import { getIsSsoUser, getLocales } from "../../selectors";
 
 const mapStateToProps = (state: State) => ({
-  user: getUser(state),
+  user: checkNotNull<User>(getUser(state)),
```
Comment (ranquild) [inline frontend/src/metabase/account/profile/containers/UserProfileApp/UserProfileApp.tsx:15]: Same here

File: frontend/src/metabase/account/profile/selectors.ts [hunk 5]
```diff
@@ -1,10 +1,18 @@
...
 
-export const getIsSsoUser = createSelector([getUser], user => {
-  return !PLUGIN_IS_PASSWORD_USER.every(predicate => predicate(user));
+import { PLUGIN_IS_PASSWORD_USER } from "metabase/plugins";
+
+import type { User } from "metabase-types/api";
+
+export const getIsSsoUser = createSelector(getUser, user => {
+  return !PLUGIN_IS_PASSWORD_USER.every(predicate =>
+    predicate(checkNotNull<User>(user)),
```
Comment (ranquild) [inline frontend/src/metabase/account/profile/selectors.ts:14]: Maybe if `user == null` return `false` immediately before `.every`?

File: frontend/src/metabase/home/homepage/containers/HomeGreeting/HomeGreeting.tsx [hunk 6]
```diff
@@ -1,10 +1,16 @@
...
-import { State } from "metabase-types/store";
+
+import type { User } from "metabase-types/api";
+import type { State } from "metabase-types/store";
+
 import HomeGreeting from "../../components/HomeGreeting";
 
 const mapStateToProps = (state: State) => ({
-  user: getUser(state),
+  user: checkNotNull<User>(getUser(state)),
```
Comment (ranquild) [inline frontend/src/metabase/home/homepage/containers/HomeGreeting/HomeGreeting.tsx:13]: Same here

**Review by ranquild — APPROVED** [review]
@kulyk small comments, please take a look

File: frontend/src/metabase/account/password/containers/UserPasswordApp/UserPasswordApp.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/account/password/containers/UserPasswordApp/UserPasswordApp.tsx:14]: Oh, nice catch, I was expecting it to be mandatory

File: frontend/src/metabase/account/profile/selectors.ts (hunk 5 above)
Comment (kulyk) [reply frontend/src/metabase/account/profile/selectors.ts]: Good call, fixed 👍

File: frontend/src/metabase/account/password/containers/UserPasswordApp/UserPasswordApp.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/account/password/containers/UserPasswordApp/UserPasswordApp.tsx]: All removed now, thanks @ranquild @WiNloSt 🙌

**Pull request merged**